server/bin/pbench-audit-server.py

```diff
@@ -0,0 +1,581 @@
+        print(
+            f"The setting for USERS in the config file is {hier.path}, but that is"
+            f" not a directory"
+        )
+        return 1
+
+    user_dirs = Path(hier.path).glob("*")
+
+    for user_path in user_dirs:
+        if user_path.is_dir():
+            hier.add_user_dir(user_path)
+        else:
+            hier.add_unexpected_objects(user_path.name)
+
+    return 0
+
+
+def check_and_dump(f, hier, ihier):
+
+    cnt = 0
+    first_check = hier.check_controller()
+    second_check = ihier.check_controller()
+    if first_check or second_check:
+        hier.header(f, "start")
+        if first_check:
+            cnt = hier.dump(f)
+        if second_check:
+            cnt += ihier.dump(f)
+        hier.header(f, "end")
+    return cnt
+
+
+def check_directory_exists(pbdir_path, pbdirname):
+    """ checks the existence of directories """
+    pbdir = Path(pbdir_path)
+    try:
+        pbdir_p = pbdir.resolve(strict=True)
+    except FileNotFoundError:
+        print(f"{_NAME_}: Bad {pbdirname}={pbdir}", file=sys.stderr)
+        return 1
+
+    if not pbdir_p.is_dir():
+        print(f"{_NAME_}: Bad {pbdirname}={pbdir}", file=sys.stderr)
+        return 1
+
+    return 0
+
+
+def main():
+    cfg_name = os.environ.get("_PBENCH_SERVER_CONFIG")
+    if not cfg_name:
+        print(
+            f"{_NAME_}: ERROR: No config file specified; set CONFIG env variable",
+            file=sys.stderr,
+        )
+        return 2
+
+    try:
+        config = PbenchServerConfig(cfg_name)
+    except BadConfig as e:
+        print(f"{_NAME_}: {e} (config file {cfg_name})", file=sys.stderr)
+        return 1
+
+    if check_directory_exists(config.ARCHIVE, "ARCHIVE") > 0:
+        return 1
+
+    if check_directory_exists(config.INCOMING, "INCOMING") > 0:
+        return 1
+
+    if check_directory_exists(config.RESULTS, "RESULTS") > 0:
+        return 1
+
+    if check_directory_exists(config.USERS, "USERS") > 0:
+        return 1
+
+    logger = get_pbench_logger(_NAME_, config)
+    init_db(config, logger)
+
+    ret = 0
+
+    try:
+        Path(config.LOGSDIR, "pbench-audit-server").mkdir()
+    except FileExistsError:
+        # directory already exists, ignore
+        pass
+    except Exception:
+        print("os.mkdir: Unable to create destination directory")
+        return 1
+    logfile = Path(config.LOGSDIR, "pbench-audit-server", "pbench-audit-server.log")
+    # TEMPORARY addition of error file for the sake of test cases
+    errorfile = Path(config.LOGSDIR, "pbench-audit-server", "pbench-audit-server.error")
+    with errorfile.open(mode="w") as f:
+        pass
+    # END
+
+    logger.info("start-{}", config.TS)
+
+    with logfile.open(mode="w") as f:
+
+        ahier = ArchiveHierarchy("archive", config.ARCHIVE, config)
+        verify_archive(ahier)
+        cnt = 0
+        if ahier.check_controller():
+            f.write(f"\nstart-{config.TS[4:]}: archive hierarchy: {config.ARCHIVE}\n")
```

Comment:
other hierarchies also depend on another hierarchy while "ARCHIVE" is independent like 'INCOMING'/'RESULTS' depend on 'CONTROLLER' and 'USER' depend on 'RESULTS' 

After processing 'ARCHIVE' we can just check whether there is anything to output. If there is any collection to be printed then we add the start and end headers and print the contents inside it.

In 'INCOMING' we check whether there is anything to output in 'CONTROLLERS' or in the 'INCOMING' hierarchy any one of them or both. if there is then we add headers. The same is the case with 'RESULTS' and 'USERS'

Reply:
The code in `header()` does this:
```
    def header(self, fp, status):
        """Adds main starting and ending messages for each Hierarchy"""
        lead_newline = "\n" if status == "start" else ""
        fp.write(
            f"{lead_newline}\n{status}-{self.config.TS[4:]}: {self.name} hierarchy: {self.path}\n"
```
The code here does this:
```
            f.write(f"\nstart-{config.TS[4:]}: archive hierarchy: {config.ARCHIVE}\n")
```
So, if you were to instead do this here
```
            ahier.header(f, "start")
```
wouldn't you get the same output (with the possible exception of an extra newline at the beginning)?

(At the end of this `if`-block, the code calls `ahier.header(f, "end")`, so it seems weird that it _doesn't_ call `ahier.header(f, "start")` at the beginning.  I assume that the issue is the extra, leading newline; and, if that's the only issue, I'd recommend using `header()` here and finding some other way to deal with the newline, like adding a separate control or parameter for it, instead of keying off of the value of the status.)

Reply:
Ok, I will try this out

Reply:
I added a new True/False argument which tells whether a new-line character should be added or not